graph/telemetry/istio/istio.go

```diff
@@ -70,91 +70,67 @@ func BuildNamespacesTrafficMap(o graph.TelemetryOptions, client *prometheus.Clie
 	// create map to aggregate traffic by protocol and response code
 	trafficMap := graph.NewTrafficMap()
-
-	requestsMetric := "istio_requests_total"
 	duration := o.Namespaces[namespace].Duration
 
-	// query prometheus for request traffic in three queries:
-	// 1) query for traffic originating from "unknown" (i.e. the internet). Unknown sources have no istio sidecar so
-	//    it is destination telemetry. Here we use destination_workload_namespace because destination telemetry
-	//    always provides the workload namespace, and because destination_service_namespace is provided from the source,
-	//    and for a request originating on a different cluster, will be set to the namespace where the service-entry is
-	//    defined, on the other cluster.
+	// HTTP/GRPC traffic
+	metric := "istio_requests_total"
 	groupBy := "source_cluster,source_workload_namespace,source_workload,source_canonical_service,source_canonical_revision,destination_cluster,destination_service_namespace,destination_service,destination_service_name,destination_workload_namespace,destination_workload,destination_canonical_service,destination_canonical_revision,request_protocol,response_code,grpc_response_status,response_flags"
-	query := fmt.Sprintf(`sum(rate(%s{reporter="destination",source_workload="unknown",destination_workload_namespace="%s"} [%vs])) by (%s)`,
-		requestsMetric,
-		namespace,
-		int(duration.Seconds()), // range duration for the query
-		groupBy)
-	unkVector := promQuery(query, time.Unix(o.QueryTime, 0), client.API())
-	populateTrafficMap(trafficMap, &unkVector, o)
-
-	// 2) query for external traffic, originating from a workload outside of the namespace.  Exclude any "unknown" source telemetry (an unusual corner
-	//	  case resulting from pod lifecycle changes).  Here use destination_service_workload to capture failed requests never reaching a dest workload.
-	reporter := "source"
-	sourceWorkloadNamespaceQuery := fmt.Sprintf(`source_workload_namespace!="%s"`, namespace)
-	query = fmt.Sprintf(`sum(rate(%s{reporter="%s",%s,source_workload!="unknown",destination_service_namespace="%s"} [%vs])) by (%s)`,
-		requestsMetric,
-		reporter,
-		sourceWorkloadNamespaceQuery,
-		namespace,
-		int(duration.Seconds()), // range duration for the query
-		groupBy)
-	extVector := promQuery(query, time.Unix(o.QueryTime, 0), client.API())
-	populateTrafficMap(trafficMap, &extVector, o)
+	idleCondition := "> 0"
+	if o.IncludeIdleEdges {
+		idleCondition = ""
+	}
 
-	// 3) query for internal traffic, originating from a workload inside of the namespace
-	query = fmt.Sprintf(`sum(rate(%s{reporter="source",source_workload_namespace="%s"} [%vs])) by (%s)`,
-		requestsMetric,
+	// 1) Incoming: query destination telemetry to capture namespace services' incoming traffic
+	query := fmt.Sprintf(`sum(rate(%s{reporter="destination",destination_service_namespace="%s"} [%vs])) by (%s) %s`,
```

Comment:
I see in some queries you are preferring `destination_service_namespace` rather than `destination_workload_namespace`. I'm wondering if this would introduce a regression: did you try the direct workload-to-workload case? I remember the graph was crashing in this case. I'm not sure if this is covered by tests.

I guess this workload-to-workload case could be catched by the 2nd query. But I think it's only partially covered, as I think a workload in "another" namespace which is calling directly a workload in "this" namespace won't be visible in the graph.

IMO, it's a corner case. But I'm not sure if this case is that common in the field so that we need to worry about it. So, it's up to you to check it.

Reply:
@israel-hdez   Well, it's a good question.  I think this may be a holdover from previous revisions of the queries.  This is destination telemetry and so it should be true that destination_workload would be set, and so it should be fine, and potentially better, to use destination_workload_namespace.

I'm fairly sure there would be no negatives to making the change, and maybe there is an advantage for the scenario you mention.  Having said that, I'm not sure it is ever true that destination_service_namespace != destination_workload_namespace when destination_workload is set.  So, I'm OK  to make the change.

The more troubling thing is something related to your scenario (a workload in another namespace which is calling directly a workload in 'this' namespace).  When thinking about it I realized that the PR currently will not capture traffic from outside the namespace that fails to reach its intended workload inside the namepace.  Put another way (using bookinfo as the target namespace):

reporter="source",
source_workload_namespace!="bookinfo", 
destination_service_namespace="bookinfo",      // actually better: destination_service=~"^.*\.bookinfo\..*$"
destination_workload_namespace="unknown", 
destination_workload="unknown"

The question is whether to add this query to the current approach or to just let users add the source namespace to the selected namespaces.

Using bookinfo I cycled down productpage to 0 pods to test this scenario.  To my surprise master also does not show the failed traffic because istio-ingressgateway does not set destination_service_namespace, only destination_service and destination_service_name.  That is why if we were to add a query we should use destination_service.   In master or in the PR I can see it fine by adding istio-system to the selected namespaces.

Any thoughts as to whether I should add the third query, another RT to Prom, for this situation?

Reply:
After thinking more on this overnight, as much as I'd prefer not to have to add a 3rd query I think it probably makes sense to add the additional query and get a complete view of incoming traffic.  In general the graph is supposed to show incoming/outgoing edge traffic from/to unselected namespacs.  And this addition, in particular, would likely show failing requests, which is valuable.

Reply:
> Having said that, I'm not sure it is ever true that destination_service_namespace != destination_workload_namespace when destination_workload is set. So, I'm OK to make the change.

I don't think services can point to workloads on a different namespace. So, I think you are right. However, I found that a service can refer to another service in another namespace. See https://stackoverflow.com/a/44329470 (it's a use-case [mentioned in official docs](https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors)) . I've never tried this. Not sure what Istio would do :-/ Anyway, we haven't heard a case. So, let's not worry about that.

> as much as I'd prefer not to have to add a 3rd query I think it probably makes sense to add the additional query and get a complete view of incoming traffic.

I agree. I was also thinking about it a little in after-work hours, and I think it's worth to add the 3rd query. I agree that the failing requests are important.

Reply:
@israel-hdez I have made another commit with both of the changes we discusses.  First, a change from destination_service_namespace to destination_workload_namespace for the destination reported telemetry.  Second, I added the new query to capture source-only telemetry for request originating outside the namespace, and failing to be serviced by a workload.  I have again removed the DNM label.   Please take a look and re-approve if it seems OK.  @pbajjuri20 the new commit will also force you to re-approve, thanks.